src/tools/stocks.py

```
@@ -1,7 +1,46 @@
 import langchain_core.tools as lcct
+import yfinance as yf
 
 
 @lcct.tool
 def get_stock_price(stock_handle: str) -> str:
-    """Mock: Return fake stock price info."""
-    return f"The stock price of {stock_handle} is $420.42 (mocked)."
+    """
+    Fetches stock information using Yahoo Finance.
+
+    Args:
+        stock_handle (str): Stock ticker (e.g., "AAPL").
+
+    Returns:
+        dict: Stock data including price and metadata.
+
+    Raises:
+        ValueError: If symbol is invalid or data is unavailable.
+        ConnectionError: If a network issue occurs.
+    """
+    try:
+        ticker = yf.Ticker(stock_handle)
+        data = ticker.info
+
+        if not data or "regularMarketPrice" not in data:
+            raise ValueError(f"Stock data not found for symbol '{stock_handle.upper()}'.")
+
+        return {
+            "symbol": stock_handle.upper(),
+            "name": data.get("shortName", "N/A"),
+            "price": data["regularMarketPrice"],
+            "previous_close": data.get("regularMarketPreviousClose", "N/A"),
+            "open": data.get("regularMarketOpen", "N/A"),
+            "high": data.get("regularMarketDayHigh", "N/A"),
+            "low": data.get("regularMarketDayLow", "N/A"),
+            "volume": data.get("regularMarketVolume", "N/A"),
+            "change": data.get("regularMarketChange", "N/A"),
+            "percent_change": f"{data.get('regularMarketChangePercent', 0):.2f}%",
+            "currency": data.get("currency", "USD"),
+            "market_state": data.get("marketState", "N/A"),
+        }
+
+    except AttributeError as e:
+        raise ValueError(f"Invalid ticker symbol '{stock_handle.upper()}' or bad response.") from e
```

I wonder if we should return this as a response instead of raise as an error

---

I think so.  A succinct returned response ("ERROR: Data for the symbol APPL was not found.  Ensure the symbol APPL is valid.") could theoretically be routed back into the LLM to solve it's own error via discovery about what stock prices might have been meant.

---

On the other hand I wonder if LangGraph handles this itself in the Tool decorator. Because it should probably handle something like the argument `a` should be int, not float.
In any case, I changed my mind, the tools should throw an error and we would handle it in the tool decorator if it is not handled yet (although I think/hope it is)